gtk/src/toga_gtk/hardware/location.py

```diff
@@ -0,0 +1,251 @@
+
+    def __init__(self, interface):
+        if Geoclue is None:
+            # CI (where coverage is enforced) must always have Geoclue available
+            # in order to perform the rest of the tests
+            raise RuntimeError(  # pragma: no cover
+                "Unable to import Geoclue. Ensure that the system package "
+                "providing Geoclue and its GTK bindings have been installed. See "
+                "https://toga.readthedocs.io/en/stable/reference/api/hardware/location.html#system-requirements "  # noqa: E501
+                "for details."
+            )
+
+        super().__init__()
+
+        self.interface = interface
+
+        #: ``Geoclue.Simple`` instance
+        self.native: Geoclue.Simple = None
+
+        #: Handle ID for ``Geoclue.Simple`` location listener
+        self.notify_location_handle: None | int = None
+
+        #: Handle ID for client active notify listener
+        self.notify_active_listener: None | int = None
+
+        self.permission_result: None | bool = None
+
+    def _start(self):
+        """Asynchronously initialize ``Geoclue.Simple``.
+
+        The act of initializing ``Geoclue.Simple`` will itself trigger
+        a permission request. As such, delay this until permission is checked.
+        """
+        self.props.state = State.STARTING
+
+        Geoclue.Simple.new(
+            self.interface.app.app_id,
+            Geoclue.AccuracyLevel.EXACT,
+            None,
+            self._new_finish,
+        )
+
+    def _new_finish(self, _, async_result):
+        try:
+            self.native = Geoclue.Simple.new_finish(async_result)
+        except GLib.Error as e:
+            if is_permissions_error(e):
+                self.props.state = State.DENIED
+            else:
+                self.props.state = State.FAILED
+
+            return
+        else:
+            self.props.state = State.READY
+
+        client = self.native.get_client()
+        # Geoclue docs indicate a client proxy is not used in sandboxed environments
+        # https://gitlab.freedesktop.org/geoclue/geoclue/-/blob/master/libgeoclue/gclue-simple.c?ref_type=heads#L978-979
+        if client:
+
+            def notify_client_active(*args):
+                if not self.native.get_client().props.active:
+                    # If the client explicitly becomes inactive, this indicates
+                    # a failure to retrieve the location
+                    # e.g., when the geoclue service is stopped on the host
+                    self.props.state = State.FAILED
+                    self._stop_tracking()
+                else:  # pragma: no cover
+                    # Not possible to meaningfully test in a platform agnostic manner
+                    self.props.state = State.READY
+
+            self.notify_active_listener = client.connect(
+                "notify::active", notify_client_active
+            )
+
+    def has_permission(self):
+        return bool(self.permission_result)
```

Comment:
If you run an example in Flatpak and use Flatseal, you can deny permissions while the app is running and already has location permission... but the XDG Location Portal doesn't send any signal to that fact, and location updates still propagate. If the app restarts, then permission does go away. I haven't found any way to revoke permissions while the app is running and after it's already requested permissions.

In a non-sandboxed environment, the state can end up in `FAILED` if the system service closes, which is handled by the listener attached to the client:

https://github.com/sarayourfriend/toga/blob/add/gtk-geoclue-geolocation/gtk/src/toga_gtk/hardware/location.py#L136-L154

A "client" is only available in non-sandboxed environments, [as mentioned in the Geoclue documentation for the method linked to in the section I've linked above](https://gitlab.freedesktop.org/geoclue/geoclue/-/blob/master/libgeoclue/gclue-simple.c?ref_type=heads#L975-976).

In that case, tracking will cease, and calls to `current_location` and `start_tracking` will fail with a `RuntimeError`.

This isn't a _permissions_ issue, though. It could be the Geoclue agent crashing, for example, and you can reproduce it by running the app outside a sandbox and running `systemctl stop geoclue`. That is why the code I linked above sets the state to `FAILED` rather than `DENIED`, and doesn't reset the permission result.